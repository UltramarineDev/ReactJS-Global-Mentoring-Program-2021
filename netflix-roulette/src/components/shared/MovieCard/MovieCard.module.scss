.cardWrapper {
  position: relative;
  padding-right: 48px;
  padding-bottom: 24px;
  max-width: 300px;
}

.nameWraper {
  width: 100%;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: flex-start;
  flex-flow: wrap;
}

.name {
  padding-bottom: 12px;
  padding-right: 36px;
  max-width: 60%;
}

.genre {
  padding-bottom: 12px;
}

.year {
  border: #ffffff;
  border-radius: 2px;
  border: solid thin #424242;
  width: 60px;
  text-align: center;
  height: 24px;
}

.dotsMenu {
  position: absolute;
  margin-left: 270px;
  margin-top: 20px;
  width: 7px;
  height: 7px;
  color: #ffffff;
}

.actions {
  position: absolute;
  min-width: 150px;
  box-shadow: 0px 8px 16px 0px rgb(0 0 0 / 20%);
  z-index: 1;
  background-color: #232323;
  margin-left: 115px;
  margin-top: 40px;
}

.actions:hover {
  cursor: pointer;
}

.dotsMenu:hover {
  cursor: pointer;
}
