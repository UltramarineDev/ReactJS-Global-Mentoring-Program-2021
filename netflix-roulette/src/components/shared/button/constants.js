export const buttonIcon = {
  ADD: "plus",
};

export const sizeStyles = {
  SMALL: 'smallSize',
  MEDIUM: 'mediumSize',
  LARGE: 'largeSize',
};
