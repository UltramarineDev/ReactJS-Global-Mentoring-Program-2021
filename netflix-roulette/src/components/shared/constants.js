export const buttonIcons = {
  ADD: 'ADD',
};

export const buttonTypes = {
  ADD: 'ADD',
  SEARCH: 'SEARCH',
  CANCEL: 'CANCEL',
};

export const backgroundImageUrl = 'https://assets.nflxext.com/ffe/siteui/vlv3/e178a4e7-4f52-4661-b2ae-41efa25dca7c/eb618394-24cd-43ae-8799-c1a76a7bcbfc/BY-en-20210222-popsignuptwoweeks-perspective_alpha_website_small.jpg';