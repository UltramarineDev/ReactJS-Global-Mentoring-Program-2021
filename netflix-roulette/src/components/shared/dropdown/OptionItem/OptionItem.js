import React from 'react';
import PropTypes from 'prop-types';

import styles from './OptionItem.module.scss';

const OptionItem = ({ label, onClick, value }) => {
  const onLinkClick = () => onClick(value);

  return (
    <a aria-label="option_item" onClick={onLinkClick} className={styles.item}>{label}</a>
  );
};

OptionItem.propTypes = {
  label: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  value: PropTypes.object.isRequired,
};

export default OptionItem;
