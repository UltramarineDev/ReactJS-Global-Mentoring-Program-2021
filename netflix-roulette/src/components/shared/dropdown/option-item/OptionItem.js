import React from 'react';
import PropTypes from 'prop-types';

import styles from './OptionItem.module.scss';

const OptionItemComponent = ({ label, onClick }) => (
  <>
    <div className={styles.item}>
      <a onClick={onClick}>{label}</a>
    </div>
  </>
);

OptionItemComponent.propTypes = {
  label: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default OptionItemComponent;
