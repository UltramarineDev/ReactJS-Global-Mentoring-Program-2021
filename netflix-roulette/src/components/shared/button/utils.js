import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
//TODO: update BUTTON_ICONS
import { BUTTON_ICONS } from '/src/components/shared/constants';

export const getButtonIcon = (icon) => {
  switch (icon) {
    case BUTTON_ICONS.ADD:
      return <FontAwesomeIcon icon="plus" />;
    default: 
      return undefined;
  }
};
