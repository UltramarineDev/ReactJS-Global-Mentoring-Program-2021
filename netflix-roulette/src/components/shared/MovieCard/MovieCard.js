import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { emptyFunc } from '/src/components/shared/constants';
import MovieImage from '../MovieImage/MovieImage';
import OptionItem from '../dropdown/OptionItem/OptionItem';
import styles from './MovieCard.module.scss';

const MovieCard = ({
  movie,
  actions, 
  onActionClick,
 }) => {
  const genre = movie ? movie.genres.join(", ") : '-';
  const year = movie ? new Date(movie.release_date).getFullYear() : '-';
  const [showOptions, setShowOptions] = useState(false);

  const handleItemClick = (id) => {
    setShowOptions(false);
    onActionClick(id, movie.id);
  };

   return (
    <div className={styles.cardWrapper}>
      <a onClick={() => setShowOptions(showOptions => !showOptions)} className={styles.dotsMenu}>
        <FontAwesomeIcon icon='ellipsis-v'/>
      </a>
      {showOptions &&
        <div className={styles.actions}>
        {actions.map(({ id, name }) => 
          <OptionItem 
            key={id}
            value={id}
            label={name}
            onClick={handleItemClick} />)}
        </div>
      }
      <a onClick={emptyFunc}>
        <MovieImage imageUrl={movie.poster_path} />
      </a>
      <div className={styles.nameWraper}>
        <div className={styles.name}>{movie.title}</div>
      <div className={styles.year}>{year}</div>
      </div>
      <div className={styles.genre}>{genre}</div>
    </div>
)};

MovieCard.propTypes = {
  movie: PropTypes.object.isRequired,
  actions: PropTypes.arrayOf(PropTypes.object.isRequired).isRequired,
  onActionClick: PropTypes.func.isRequired,
};

export default MovieCard;
