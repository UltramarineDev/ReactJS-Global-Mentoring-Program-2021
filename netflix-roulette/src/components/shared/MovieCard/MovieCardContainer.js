import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import MovieCard from './MovieCard';

class MovieCardContainer extends PureComponent {
  static propTypes = {
    movie: PropTypes.object.isRequired,
    onActionClick: PropTypes.func.isRequired,
    actions: PropTypes.arrayOf(PropTypes.object.isRequired).isRequired,
  };

   state = {
    showOptions: false,
  }

  handleActionsClick = () => this.setState(({ showOptions }) => { 
    return { showOptions: !showOptions } 
  });

  handleMovieClick = () => {};

  render() {
    const { movie, onActionClick, actions } = this.props;
    const { showOptions } = this.state;

    const genre = movie ? movie.genres.join(", ") : '-';
    const year = movie ? new Date(movie.release_date).getFullYear() : '-';
    return (
      <MovieCard
        movie={movie}
        year={year}
        genre={genre}
        onActionsClick={this.handleActionsClick}
        showOptions={showOptions}
        onActionClick={onActionClick}
        actions={actions}
        onMovieClick={this.handleMovieClick}
      />
    );
  }
};

export default MovieCardContainer;
