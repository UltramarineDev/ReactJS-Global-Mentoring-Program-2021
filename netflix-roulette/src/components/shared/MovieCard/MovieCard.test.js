import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { fireEvent, screen, render, queryByAttribute, cleanup } from '@testing-library/react';

import { getMovieAction } from 'actions';

import { getOptions } from '../MoviesList/utils';
import MovieCard from './MovieCard';
import { movieActions, movieActionLabels } from '../MoviesList/constants';
import OptionItem from '../dropdown/OptionItem/OptionItem';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch,
}));

jest.mock('actions', () => ({
  getMovieAction: jest.fn() }
));

jest.mock('../dropdown/OptionItem/OptionItem');

describe('MovieCard tests', () => {
  afterEach(cleanup);
  test('renders MovieCard', () => {
    const onActionClick = jest.fn();
    const movie = {};
    const actions = getOptions(movieActions, movieActionLabels);
    const component = renderer.create(
      <MemoryRouter>
        <MovieCard onActionClick={onActionClick} movie={movie} actions={actions} />
      </MemoryRouter>,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders MovieCard with showOptions', () => {
    const onActionClick = jest.fn();
    const movie = {};
    const actions = getOptions(movieActions, movieActionLabels);
    const dom = render(
      <MemoryRouter>
        <MovieCard onActionClick={onActionClick} movie={movie} actions={actions} />
      </MemoryRouter>,
    );
    const getById = queryByAttribute.bind(null, 'id');
    const link = getById(dom.container, 'set_options');
    OptionItem.mockImplementation(() => <div>OptionItem</div>);

    fireEvent.click(link);
    expect(screen.getAllByText('OptionItem')).toHaveLength(2);
  });

  test('renders MovieCard with getMovieAction', () => {
    const onActionClick = jest.fn();
    const movie = { id: 1 };
    const actions = getOptions(movieActions, movieActionLabels);
    const dom = render(
      <MemoryRouter>
        <MovieCard onActionClick={onActionClick} movie={movie} actions={actions} />
      </MemoryRouter>,
    );
    const getById = queryByAttribute.bind(null, 'id');
    const link = getById(dom.container, 'movie_link');

    fireEvent.click(link);
    expect(getMovieAction).toBeCalled();
  });
});
