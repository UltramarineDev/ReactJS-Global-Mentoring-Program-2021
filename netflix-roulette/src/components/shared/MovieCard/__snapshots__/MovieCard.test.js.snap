// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard tests renders MovieCard 1`] = `
<div
  className="cardWrapper"
>
  <a
    className="dotsMenu"
    id="set_options"
    onClick={[Function]}
  />
  <a
    href="/movie/undefined"
    id="movie_link"
    onClick={[Function]}
  >
    <div
      className="noImageWrapper"
    />
  </a>
  <div
    className="nameWraper"
  >
    <div
      className="name"
    />
    <div
      className="year"
    >
      NaN
    </div>
  </div>
  <div
    className="genre"
  >
    -
  </div>
</div>
`;
